File: Longest Increasing Subsequence/Python/Recursive/Lis.py
# Longest Increasing Subsequence (LIS) problem using recursion

# calculate max(a,b)
def max(a,b):
    if a >= b:
        return a
    return b

def localLis(array, dim, i, currentMax):
    # base case
    if i == dim:
        return 0
    # recursive case 1: array[i] >= currentMax
    if array[i] >= currentMax:
        return 1 + localLis(array, dim, i+1, array[i])
    # recursive case 2: array[i] < max : ignore current value
    return localLis(array, dim, i+1, currentMax)

# Calculate LIS
def lis(array, dim, i, maximum):
    # base case
    if i >= dim-1:
        return maximum
    # recursive case
    return max(localLis(array, dim, i, maximum), lis(array, dim, i+1, maximum))

value = raw_input("Enter value:\n")
array = [int(x) for x in value.split(" ")]
result = lis(array, len(array), 0, 1)
print("Longest Increasing Subsequence (LIS) for {0} has length: {1}".format(value, result))

File: Longest Increasing Subsequence/Java/Dynamic-Programming/Lis.java
import java.util.Arrays;
import java.util.Scanner;
/**
 * Implementation of Longest Increasing Subsequence (LIS) 
 * problem using dynamic programming
 */

class Lis {

    private static int findMax(int[] lis) {
        return Arrays.stream(lis).max().getAsInt();
    }

    private static int lis(int[] array) {
        int[] lis = new int[array.length];
        // Lis for length = 1 is 1
        for (int i=0; i<array.length; i++) {
            lis[i] = 1;
        }
        // foreach dim, update lis
        for (int i=1; i<array.length; i++) {
            for (int j=0; j<i; j++) {
                if (array[i] > array[j] && lis[i] < lis[j] + 1) {
                    lis[i] = 1 + lis[j];
                }
            }
        }
        return findMax(lis);
    }

    static int[] parseIntArray(String s) {
        return Arrays.stream(s.split(" "))
            .mapToInt(Integer::parseInt)
            .toArray();
    }

    public static void main(String args[]) {
        Scanner scanner = new Scanner(System.in);
        System.out.println("Insert number's sequence:");
        String input = scanner.nextLine();
        scanner.close();
        int[] array = parseIntArray(input);
        //Calculate result
        String result = String.format("Longest Increasing Subsequence (LIS) for %s has length: %s", input, lis(array));
        System.out.println(result);
    }
}